File: ios/ExpoYandexMapKitPolygonView.swift
import ExpoModulesCore
import YandexMapsMobile

// One `<Polygon>`. Drives a MapKit YMKPolygonMapObject; created lazily once the collection and an
// outer ring of at least three points are available.
class ExpoYandexMapKitPolygonView: ExpoView, MapObjectChild {
  let onShapePress = EventDispatcher()

  private var collection: YMKMapObjectCollection?
  private var mapObject: YMKPolygonMapObject?
  private var tapListener: PolygonTapListener?
  private var points: [YMKPoint] = []
  private var innerRings: [[YMKPoint]] = []
  private var fillColor: UIColor?
  private var strokeColor: UIColor?
  private var strokeWidth: Float?
  private var zIndexValue: Float = 0
  private var handled = false

  func setPoints(_ value: [PointRecord]) {
    points = value.map { YMKPoint(latitude: $0.latitude, longitude: $0.longitude) }
    updateGeometry()
  }

  func setInnerRings(_ value: [[PointRecord]]) {
    innerRings = value.map { ring in ring.map { YMKPoint(latitude: $0.latitude, longitude: $0.longitude) } }
    updateGeometry()
  }

  func setFillColor(_ value: UIColor?) {
    fillColor = value
    updateStyle()
  }

  func setStrokeColor(_ value: UIColor?) {
    strokeColor = value
    updateStyle()
  }

  func setStrokeWidth(_ value: Double) {
    strokeWidth = Float(value)
    updateStyle()
  }

  func setZIndexValue(_ value: Double) {
    zIndexValue = Float(value)
    updateStyle()
  }

  func setHandled(_ value: Bool) {
    handled = value
  }

  func attachToMap(_ collection: YMKMapObjectCollection) {
    self.collection = collection
    createIfReady()
  }

  func detachFromMap(_ collection: YMKMapObjectCollection) {
    if let obj = mapObject {
      collection.remove(with: obj)
    }
    mapObject = nil
    tapListener = nil
    self.collection = nil
  }

  var isAttachedToMap: Bool { collection != nil }

  private func buildPolygon() -> YMKPolygon {
    YMKPolygon(
      outerRing: YMKLinearRing(points: points),
      innerRings: innerRings.map { YMKLinearRing(points: $0) })
  }

  private func createIfReady() {
    guard mapObject == nil, let collection = collection, points.count >= 3 else {
      return
    }
    let obj = collection.addPolygon(with: buildPolygon())
    mapObject = obj
    let listener = PolygonTapListener(view: self)
    tapListener = listener
    obj.addTapListener(with: listener)
    updateStyle()
  }

  private func updateGeometry() {
    if let obj = mapObject, obj.isValid, points.count >= 3 {
      obj.geometry = buildPolygon()
    } else {
      createIfReady()
    }
  }

  private func updateStyle() {
    guard let obj = mapObject, obj.isValid else {
      return
    }
    obj.zIndex = zIndexValue
    if let strokeWidth = strokeWidth {
      obj.strokeWidth = strokeWidth
    }
    if let strokeColor = strokeColor {
      obj.strokeColor = strokeColor
    }
    if let fillColor = fillColor {
      obj.fillColor = fillColor
    }
  }

  fileprivate func handleTap(_ point: YMKPoint) -> Bool {
    onShapePress(["point": ["latitude": point.latitude, "longitude": point.longitude]])
    return handled
  }
}

private final class PolygonTapListener: NSObject, YMKMapObjectTapListener {
  private weak var view: ExpoYandexMapKitPolygonView?

  init(view: ExpoYandexMapKitPolygonView) {
    self.view = view
  }

  func onMapObjectTap(with mapObject: YMKMapObject, point: YMKPoint) -> Bool {
    return view?.handleTap(point) ?? false
  }
}

File: android/src/main/java/expo/modules/yandexmapkit/ExpoYandexMapKitPolygonView.kt
package expo.modules.yandexmapkit

import android.content.Context
import com.yandex.mapkit.geometry.LinearRing
import com.yandex.mapkit.geometry.Point
import com.yandex.mapkit.geometry.Polygon
import com.yandex.mapkit.map.MapObject
import com.yandex.mapkit.map.MapObjectCollection
import com.yandex.mapkit.map.MapObjectTapListener
import com.yandex.mapkit.map.PolygonMapObject
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import java.lang.ref.WeakReference

// One `<Polygon>`. Drives a MapKit PolygonMapObject; created lazily once the collection and an
// outer ring of at least three points are available.
class ExpoYandexMapKitPolygonView(context: Context, appContext: AppContext) :
  ExpoView(context, appContext), MapObjectTapListener, MapObjectChild {
  private val onPress by EventDispatcher<Map<String, Any?>>()

  private var collection: MapObjectCollection? = null
  private var mapObject: PolygonMapObject? = null
  private var points: List<Point> = emptyList()
  private var innerRings: List<List<Point>> = emptyList()
  private var fillColor: Int? = null
  private var strokeColor: Int? = null
  private var strokeWidth: Float? = null
  private var zIndexValue = 0f
  private var handled = false

  internal fun setPoints(value: List<Point>) {
    points = value
    updateGeometry()
  }

  internal fun setInnerRings(value: List<List<Point>>) {
    innerRings = value
    updateGeometry()
  }

  internal fun setFillColor(value: Int?) {
    fillColor = value
    updateStyle()
  }

  internal fun setStrokeColor(value: Int?) {
    strokeColor = value
    updateStyle()
  }

  internal fun setStrokeWidth(value: Float) {
    strokeWidth = value
    updateStyle()
  }

  internal fun setZIndexValue(value: Float) {
    zIndexValue = value
    updateStyle()
  }

  internal fun setHandled(value: Boolean) {
    handled = value
  }

  override fun attachToMap(collection: MapObjectCollection) {
    this.collection = collection
    createIfReady()
  }

  override fun detachFromMap(collection: MapObjectCollection) {
    mapObject?.let { collection.remove(it) }
    mapObject = null
    this.collection = null
  }

  override val isAttachedToMap: Boolean
    get() = collection != null

  private fun buildPolygon() = Polygon(LinearRing(points), innerRings.map { LinearRing(it) })

  private fun createIfReady() {
    if (mapObject != null) {
      return
    }
    val collection = collection ?: return
    if (points.size < 3) {
      return
    }
    val obj = collection.addPolygon(buildPolygon())
    mapObject = obj
    obj.addTapListener(WeakReference(this))
    updateStyle()
  }

  private fun updateGeometry() {
    val obj = mapObject
    if (obj != null && obj.isValid && points.size >= 3) {
      obj.geometry = buildPolygon()
    } else {
      createIfReady()
    }
  }

  private fun updateStyle() {
    val obj = mapObject ?: return
    if (!obj.isValid) {
      return
    }
    obj.zIndex = zIndexValue
    strokeWidth?.let { obj.strokeWidth = it }
    strokeColor?.let { obj.strokeColor = it }
    fillColor?.let { obj.fillColor = it }
  }

  override fun onMapObjectTap(mapObject: MapObject, point: Point): Boolean {
    onPress(mapOf("point" to mapOf("latitude" to point.latitude, "longitude" to point.longitude)))
    return handled
  }
}
